core/src/sumeragi/main_loop.rs

```diff
@@ -430,325 +435,378 @@ impl Sumeragi {
         message: BlockMessage,
         state: &'state State,
         voting_block: &mut Option<VotingBlock<'state>>,
-        current_view_change_index: u64,
+        view_change_index: usize,
         genesis_public_key: &PublicKey,
-        voting_signatures: &mut Vec<SignatureOf<BlockPayload>>,
+        voting_signatures: &mut Vec<BlockSignature>,
+        #[cfg_attr(not(debug_assertions), allow(unused_variables))] is_genesis_peer: bool,
     ) {
-        let current_topology = &self.current_topology;
-        let role = current_topology.role(&self.peer_id);
-        let addr = &self.peer_id.address;
-
         #[allow(clippy::suspicious_operation_groupings)]
-        match (message, role) {
+        match (message, self.role()) {
             (BlockMessage::BlockSyncUpdate(BlockSyncUpdate { block }), _) => {
-                let block_hash = block.hash();
-                info!(%addr, %role, block=%block_hash, "Block sync update received");
+                info!(
+                    peer_id=%self.peer_id,
+                    role=%self.role(),
+                    block=%block.hash(),
+                    "Block sync update received"
+                );
+
+                // FIX: Release writer before handling block sync
+                let curr_block = voting_block.take();
```

Comment:
I have this issue in 4 places. It happens either on `BlockSyncUpdate` or `BlockCreated`. At the time of receiving a new block it's possible that the block is invalid and I would like to keep the old block while validating the new one. However, this creates deadlock as I have not released previous writer

@Erigara do you have some suggestions?

Reply:
Unfortunately it’s current limitation of our state layout. 
That it’s single writer and locks state for duration of block.

When i’ve designed state i made some decisions for making optimisations for happy path (block committed) without problems. 
We might reconsider some of this decisions if necessary.

Maybe we can be smarter about when to reject incoming block? Like checking signatures, height, hashes this checks doesn’t need ability to modify state.

Reply:
well, I'll just drop the current voting block in this PR as it was before and open an issue for this

Reply:
opened #4643